servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/SequentialExecutor.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright © 2023 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.loadbalancer;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A concurrency primitive for providing thread safety without using locks.
+ *
+ * A {@link SequentialExecutor} is queue of tasks that are executed one at a time in the order they were
+ * received. This provides a way to serialize work between threads without needing to use locks which can
+ * result in thread contention and thread deadlock scenarios.
+ */
+final class SequentialExecutor implements Executor {
```

Comment:
There is plenty of room for opinion about how this should be implemented but the fundamental point is that it's a task queue that executes tasks one at a time and in the order they were submitted.

Alternative ways to implement a queue:
* ConcurrentLinkedList with a 'running' flag
* Processor with a `.forEach(Runnable::run)` call
* ...

Reply:
Absolutely. I wanted to wait until I have buy-in but I imagine it's a blocker to not have tests.

Reply:
I added some tests in https://github.com/apple/servicetalk/pull/2768/commits/964dbc62e3aca1275b5ebca38ec126e9e0ce6028. Feedback welcome.